doc/advanced-testing-mini-course/README.md

```diff
@@ -83,22 +92,96 @@ step m cmd = case cmd of
 exec :: Counter -> Command -> IO Response
 exec = undefined
 
-genCommands :: Model -> Gen [Command]
-genCommands = undefined
+newtype Program = Program [Command]
+
+genProgram :: Model -> Gen Program
+genProgram = undefined
 
 prop_counter :: Property
-prop_counter = forAll genCommands $ \cmds -> do
+prop_counter = forAll genProgram $ \(Program cmds) -> do
   undefined
 ```
 
+* Excerises
+
+1. Implement shrinking for programs.
+
+2. Collect timing information about how long each command takes to execute on
+   average.
+
+* See also
+
+  - Why state machines over other forms of specifications?
+    + Executable (we will use this later)
+    + Gurevich's generalisation of the Church-Turing thesis
+    + Already heavily used in distributed systems
+
```

Comment:
We should probably also have an exercise of
-  finding a bug
-  make a regression test
-  fix bug
To make sure that we teach early about importance of regression tests.

Reply:
Would be nice if we did all that together while live coding for some contrived bug, but then left an other bug as exercise (that perhaps needs generator tweaking to appear, or at least bigger `--quickcheck-size`)...

Reply:
Or perhaps the exercise should be to test a completely different SUT?

Reply:
Yeah the exercise might be a different SUT